src/frontends/lean/info_manager.cpp

```diff
@@ -112,9 +120,68 @@ void vm_obj_format_info::report(io_state_stream const & ios, json & record) cons
 }
 #endif
 
+json widget_info::to_json() const {
+    vdom * vd = const_cast<vdom*>(&m_vdom);
+    return vd->to_json();
+}
+
+void widget_info::report(io_state_stream const & ios, json & record) const {
+    if (!get_global_module_mgr()->get_report_widgets()) { return; }
+    vm_state S(m_env, ios.get_options());
+    scope_vm_state scope(S);
+    record["widget"]["html"] = to_json();
+}
+
+void widget_info::update(io_state_stream const & ios, json const & message, json & record) const {
```

Comment:
```suggestion
void widget_info::update(io_state_stream const & ios, json const & message, json & record) {
```

You're already const-casting in the other `update` method.